tools/codegen/api/translate.py

```diff
@@ -115,38 +115,38 @@ def unsat(goal: CType) -> NoReturn:
     # conversions (e.g., "T a" is OK for "const T& a").  So all of the
     # existing rules in this function simply try to solve immediately,
     # and bail if things don't work out.
-    def solve(goal: CType, *, direct: bool) -> str:
-        def direct_solve(goal: CType) -> str:
+    def solve(goal: NamedCType, *, direct: bool) -> str:
+        def direct_solve(goal: NamedCType) -> str:
             return solve(goal, direct=True)
 
         if goal in ctx:
             # Trivial
             return ctx[goal]
 
         # const & is satisfied with mutable &
-        if isinstance(goal, ConstRefCType):
+        if isinstance(goal.type, ConstRefCType):
             try:
                 # WARNING: not strictly decreasing; be careful not
                 # to add a direct conversion that goes satisfies
                 # mutable& with const&
-                return solve(MutRefCType(goal.elem), direct=direct)
+                return solve(NamedCType(goal.name, MutRefCType(goal.type.elem)), direct=direct)
```

Comment:
This kind of thing is a downside of separating out `CType` and `NamedCType` - which Ed already pointed out. The "name" in `NamedCType` is only attached to the top-level CType. But it does make comparing C++ types easier :)